### ISSUE TITLE
Dealing with XMLs that contains invalid character references

### ISSUE THREAD

**Issue by zudov** [snoyberg/xml#119]
Since #112 `xml-conduit` doesn't allow character references that represent illegal XML characters by failing a parser.

The service that we deal with unfortunately returns some of those characters (NULL and some ASCII controls) in a few places. Previously we parsed the XML and then removed the bad characters from the extracted text content (after we found out that they are there). It's better and safer if the parser yells about these characters during parsing (thus forcing us to deal with potential problem), however as it-is there is no easy way to deal with such "slightly invalid" XMLs which breaks our use-case™.

I open an issue for two reasons:

1. It seems that the error message with which the parser fails isn't as descriptive as it was intended:
   ```haskell
   *Text.XML> parseText def "<tag>&#x0;</tag>"
   Left (ParseError {errorContexts = [], errorMessage = "Failed reading: takeWhile1", errorPosition = 1:6 (5)})
   ```
   While the [parsing code](https://github.com/snoyberg/xml/pull/112/files#diff-54706cc7ebafb4666b8b1bec67419378R589) clearly tries to provide a good error message:
   
   ```haskell
   parseHexCharRef = do
     void $ string "#x"
     n <- AT.hexadecimal
     case toValidXmlChar n of
       Nothing -> fail "Invalid character from hexadecimal character reference."
       Just c -> return $ ContentText $ T.singleton c
   ```

2. Let's discuss what would be a good way to allow dealing with XML that contains these characters.
   We were thinking that it would be nice to have an option which would cause the parser to skip
   these character references instead of failing. The UX would then be like:
   
   - try feeding the funky XML input into the parser
   - get an "Invalid character from hexadecimal character reference" error
   - figure out what can be done about it and enable an "ignore invalid characters" option
   
   Our current (hopefully temporary) solution is to pass the input string through the following conduit:
   
   ```haskell
   -- | Removes invalid hexadecimal XML character references from the stream.
   skipInvalidCharRefs :: MonadThrow m => Conduit ByteString m ByteString
   skipInvalidCharRefs =
       Atto.conduitParser token .| mapC snd .| filterC (not . ByteString.null)
     where
       token :: Atto.Parser ByteString
       token =
         Atto.takeWhile1 (/= '&')
           <|>
         "" <$ mfilter (not . isValidRef) (hexRef <|> decRef)
           <|>
         "&"
   
       hexRef :: Atto.Parser Int
       hexRef = "&#x" *> Atto.hexadecimal <* ";"
   
       decRef :: Atto.Parser Int
       decRef = "&#" *> Atto.decimal <* ";"
   
       isValidRef :: Int -> Bool
       isValidRef n =
         any (\(lb, ub) -> lb <= n && n <= ub) validRanges
   
       validRanges :: [(Int, Int)]
       validRanges =
         [ (0x9, 0xA)
         , (0xD, 0xD)
         , (0x20, 0xD7FF)
         , (0xE000, 0xFFFD)
         , (0x10000, 0x10FFFF)
         ]
   ```

These changes can be compatible with no need for major releases. Let's consider this and I can make the changes.

**Comment by k0ral** [comment]
1/ From what I tested, the misleading error message is due to the use of `attoparsec` behind the scenes, which doesn't seem to track errors accurately when using `<|>`. I suppose this is intended for performance reasons, I'm not sure what we can do about it.

2/ Rather than ignoring invalid hexadecimal characters, I would suggest we add a `type DecodeHexChar :: Int -> Maybe Content`, similar to the existing `DecodeEntities`, so that users can customize how invalid hexadecimal characters should be interpreted (`Nothing` should lead to a parsing error). What do you think ?

**Comment by zudov** [comment]
:+1: to having a user-provided reference handling function. Definitely better than hard-coding a specific behavior. Might make sense to pass it as a `Char` instead of `Int` (more descriptive type) on the other hand `Int` is how it appears in the XML, so I am on the fence on whether it should be `Int` or `Char`.

**Comment by k0ral** [comment]
Available in [1.7.1.0](https://hackage.haskell.org/package/xml-conduit-1.7.1.0).
